public/app/features/alerting/unified/components/rules/state-history/common.ts

```diff
@@ -37,3 +44,99 @@ export function extractCommonLabels(labels: Label[][]): Label[] {
+  const lines = stateHistory?.data?.values[1] ?? [];
+
+  const logRecords = timestamps.reduce((acc: LogRecord[], timestamp: number, index: number) => {
+    const line = lines[index];
+    // values property can be undefined for some instance states (e.g. NoData)
+    if (isLine(line)) {
+      acc.push({ timestamp, line });
+    }
+
+    return acc;
+  }, []);
+
+  // group all records by alert instance (unique set of labels)
+  const logRecordsByInstance = groupBy(logRecords, (record: LogRecord) => {
+    return JSON.stringify(record.line.labels);
+  });
+
+  return { logRecordsByInstance, logRecords };
+};
+
+export function getRuleHistoryRecordsForPanel(stateHistory?: DataFrameJSON) {
+  if (!stateHistory) {
+    return { dataFrames: [] };
+  }
+  const theme = config.theme2;
+
+  const { logRecordsByInstance } = getLogRecordsByInstances(stateHistory);
+
+  const groupedLines = Object.entries(logRecordsByInstance);
+
+  const dataFrames: DataFrame[] = groupedLines.map<DataFrame>(([key, records]) => {
+    return logRecordsToDataFrameForPanel(key, records, theme);
+  });
+
+  return {
+    dataFrames,
+  };
+}
+
+export const getHistoryImplementation = () => {
+  // can be "loki", "multiple" or "annotations"
+  const stateHistoryBackend = config.unifiedAlerting.alertStateHistoryBackend;
+  // can be "loki" or "annotations"
+  const stateHistoryPrimary = config.unifiedAlerting.alertStateHistoryPrimary;
+
+  // if "loki" is either the backend or the primary, show the new state history implementation
+  const usingNewAlertStateHistory = [stateHistoryBackend, stateHistoryPrimary].some(
+    (implementation) => implementation === StateHistoryImplementation.Loki
+  );
+  const implementation = usingNewAlertStateHistory
+    ? StateHistoryImplementation.Loki
+    : StateHistoryImplementation.Annotations;
+  return implementation;
+};
+
+export const updatePanelDataWithASHFromLoki = async (panelDataProcessed: PanelData) => {
+  //--- check if alert state history uses Loki as implementation, if so, fetch data from Loki state history and concat it to annotations
+  const historyImplementation = getHistoryImplementation();
+  const usingLokiAsImplementation = historyImplementation === StateHistoryImplementation.Loki;
+
+  if (
+    !usingLokiAsImplementation ||
+    !panelDataProcessed.alertState?.dashboardId ||
+    !panelDataProcessed.alertState?.panelId
+  ) {
+    return panelDataProcessed;
+  }
+
+  try {
+    // fetch data from Loki state history
+    let annotationsWithHistory = await getBackendSrv().get('/api/v1/rules/history', {
+      panelID: panelDataProcessed.request?.panelId,
+      dashboardUID: panelDataProcessed.request?.dashboardUID,
+      from: panelDataProcessed.timeRange.from.unix(),
+      to: panelDataProcessed.timeRange.to.unix(),
+      limit: 250,
+    });
+    const records = getRuleHistoryRecordsForPanel(annotationsWithHistory);
+    const clonedPanel = cloneDeep(panelDataProcessed);
```

Comment:
nit: prefer `produce` here and should be a bit more readable with an `if` statement

```typescript
const newPanelData = produce(panelDataProcessed, draft => {
  if (Array.isArray(draft.annotations)) {
    draft.annotations = draft.annotations.concat(records.dataFrames)
  }

  return draft;
});
```